### PR TITLE
New Resource: `azurerm_mariadb_firewall_rule` (resolves #3608)

### LINKED ISSUE #3608: Add missing azurerm_mariadb_firewall_rule
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add a new azurerm_mariadb_firewall_rule resource, following the model of azurerm_mysql_firewall_rule

### New or Affected Resource(s)

* azurerm_mariadb_firewall_rule

### Potential Terraform Configuration

```hcl
resource "azurerm_mariadb_firewall_rule" "mariadb_firewall_azure" {
   name                = "azure"
   resource_group_name = var.resource_group_name
   server_name         = "${azurerm_mariadb_server.mariadb_server.name}"
   start_ip_address    = "0.0.0.0"
   end_ip_address      = "0.0.0.0"
}
```

### References

* https://github.com/Azure/azure-sdk-for-go/blob/master/services/mariadb/mgmt/2018-06-01/mariadb/firewallrules.go
* https://docs.microsoft.com/en-us/cli/azure/mariadb/server/firewall-rule?view=azure-cli-latest

### PR DESCRIPTION
**Pull request by armbraggins**

### CODE REVIEW

**Comment by katbyte** [comment]
Also a test is failing:
```

------- Stdout: -------
=== RUN   TestAccAzureRMMariaDbFirewallRule_basic
=== PAUSE TestAccAzureRMMariaDbFirewallRule_basic
=== CONT  TestAccAzureRMMariaDbFirewallRule_basic
--- FAIL: TestAccAzureRMMariaDbFirewallRule_basic (0.10s)
    testing.go:568: Step 0 error: config is invalid: expected version to be one of [10.2], got 5.6
```

File: azurerm/resource_arm_mariadb_firewall_rule.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
```
[the remark this answers is not in the document]
Comment (armbraggins) [reply azurerm/resource_arm_mariadb_firewall_rule.go]: Neither postgresql_firewall_rule nor mysql_firewall_rule do. Is there an example I could follow?

File: azurerm/resource_arm_mariadb_firewall_rule.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"server_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
```
[the remark this answers is not in the document]
Comment (armbraggins) [reply azurerm/resource_arm_mariadb_firewall_rule.go]: Neither postgresql_firewall_rule nor mysql_firewall_rule do. Is there an example I could follow?

File: azurerm/resource_arm_mariadb_firewall_rule.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+			"start_ip_address": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
[the remark this answers is not in the document]
Comment (armbraggins) [reply azurerm/resource_arm_mariadb_firewall_rule.go]: Neither postgresql_firewall_rule nor mysql_firewall_rule do, but I'll try adding
"ValidateFunc: validate.IPv4Address,"

**Comment by armbraggins** [comment]
While this worked for me when I first tried it, I'm currently seeing
"mariadb.FirewallRulesClient#CreateOrUpdate: Failure preparing request: StatusCode=0 -- Original Error: autorest: No scheme detected in URL"

**Comment by armbraggins** [comment]
> While this worked for me when I first tried it, I'm currently seeing
> "mariadb.FirewallRulesClient#CreateOrUpdate: Failure preparing request: StatusCode=0 -- Original Error: autorest: No scheme detected in URL"

I'm still hoping to learn about debugging Terraform providers as a background task, but for now I'm using local-exec and az cli as a workaround,

File: azurerm/config.go [hunk 4]
```diff
@@ -189,6 +189,7 @@ type ArmClient struct {
 	// Databases
 	mariadbDatabasesClient                   mariadb.DatabasesClient
 	mariadbServersClient                     mariadb.ServersClient
+	mariadbFirewallRulesClient               mariadb.FirewallRulesClient
```
Comment (katbyte) [inline azurerm/config.go]: the reason you are seeing that error is you haven't registered the client. You'll have to add something like:
```
	c.BackendClient = apimanagement.NewBackendClientWithBaseURI(o.ResourceManagerEndpoint, o.SubscriptionId)
	o.ConfigureClient(&c.BackendClient.Client, o.ResourceManagerAuthorizer)
```

in the appropriate area.

File: azurerm/resource_arm_mariadb_firewall_rule.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (katbyte) [reply azurerm/resource_arm_mariadb_firewall_rule.go]: Look at `ValidateMsSqlDatabaseName` in helpers/azure/mssql. you should create a new file there and add mariadb equivalents.

File: azurerm/resource_arm_mariadb_firewall_rule.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (katbyte) [reply azurerm/resource_arm_mariadb_firewall_rule.go]: as mentioned above

**Comment by mbfrahry** [comment]
Hey @armbraggins, it looks like this resource got merged in from another contributor. Thank you so much for taking the time to take a crack at this but unfortunately we're going to have to close it down.

**Pull request closed without merging**